### ISSUE TITLE
Provide a way to set properties with Integer (not Long) values in Cypher

### ISSUE THREAD

**Issue by codeseedr** [neo4j/neo4j#7652]
API allows you to specify either Integer or Long as an integral type for your property. Unfortunately Cypher is limited to Long only. This poses an issue if you want to set your property through Cypher and then read it through a program which expects the property to be of Integer type. On such occasion you will get a ClassCastException.

Perhaps a cast like (INTEGER) would do. Unfortunately, since TOINT() function obviously in reality returns a Long, there would be a bit of weirdness like "(INTEGER) TOINT("123")", if someone wanted to convert a String to Integer.

**Comment by spacecowboy** [comment]
When you say 

> and then read it through a program which expects the property to be of Integer type

What do you mean exactly? Why can't your program do `(int) x`?

**Comment by codeseedr** [comment]
If neo4j is accessed by a 3rd party library, I can't program it that way... And I need to massage data through Cypher.

Anyway, I have a feeling that Cypher is handicapped. It seems very weird to me not to be able to use a certain type through query language (and it's nothing sophisticated really, just the good, old Integer). I understand implementation with a single integral type is easier and if you are not willing to do that (because of numerous factors) you may also look at the issue in terms of consistency. For me neo4j should either support Long only or both types on all fronts.

**Comment by spacecowboy** [comment]
Reading your initial message again

> API allows you to specify either Integer or Long as an integral type for your property. 

You mean the embedded API here, right?

**Comment by codeseedr** [comment]
Correct.

**Comment by codeseedr** [comment]
Thanks for recognising it as a valid feature. Though one can argue it's a bug as well. Anyway, thanks. :)

**Comment by yxdunc** [comment]
Hello,
Doing a toFloat with an integer value: `toFLoat(42)` seems to return a normal integer.

Works well for me :)

**Comment by codeseedr** [comment]
Haha, that's a bug actually.

**Comment by jexp** [comment]
Cypher itself and Bolt only know whole numbers and floating point numbers there is intentionally no small-type support as that's an implementation detail.

That things still work in the java API is a historic development issue.

So even if you set a property to int in java if you update it with plain cypher it'll turn into a long.

If you want to use the value as Integer in java do: `((Number)result.get("name")).intValue()`

**Comment by codeseedr** [comment]
So either remove support for int entirely or provide means to use it within Cypher. Currently neo4j is somewhere in the middle which causes headaches if I want to alter data from Cypher and that data is read & written by an external program coded to use int properties.

**Comment by jexp** [comment]
Yes we're aware of that. We can't just remove the Java API /other types for backward compatibility. 

There will be a proper deprecation and removal process to make sure we don't break existing apps and projects.  

If you're aware of the fact for your own app where you mix cypher and embedded you can accommodate for it. 

Von meinem iPhone gesendet

**Comment by MatthewDavidCampbell** [comment]
+1 on numerical type support.  Posted a [stackoverflow](https://stackoverflow.com/questions/45997135/lack-of-certain-neo4j-scalar-functions) before seeing this issue.  Doesn't a limitation in Cypher to long and float effect the property store size leading to bloated graphs on disk?

**Comment by codeseedr** [comment]
I tried writing a custom function that would return Integer but that proved futile due to:
```
org.neo4j.kernel.api.exceptions.ProcedureException: Don't know how to map `java.lang.Integer` to the Neo4j Type System.
Please refer to to the documentation for full details.
For your reference, known types are: [boolean, byte[], double, java.lang.Boolean, java.lang.Double, java.lang.Long, java.lang.Number, java.lang.Object, java.lang.String, java.util.List, java.util.Map, long, org.neo4j.graphdb.Node, org.neo4j.graphdb.Path, org.neo4j.graphdb.Relationship, org.neo4j.graphdb.spatial.Geometry, org.neo4j.graphdb.spatial.Point]
```

**Comment by jexp** [comment]
Why would you need that?
It's just a whole number

**Comment by codeseedr** [comment]
Because of the very issue this ticket describes. There's no cypher-level support for integer data type which you can use when using programmatic API. There's no way to update an integer property through cypher that would not get it promoted to long, breaking the code that tries to access it as integer.

**Comment by cashlalala** [comment]
So, until today, there's still no way to set a integer value to a property via cypher?

**Comment by codeseedr** [comment]
Correct, no solution to this day. Unless anybody can prove otherwise.

**Comment by technige** [comment]
@codeseedr As of Neo4j 3.0, which introduced Bolt and the official drivers, back in 2016, the core type system for the primary surface can be summarised as follows:

- Null
- Boolean
- Integer (signed 64-bit integer)
- Float (double precision floating point number)
- String (Unicode)
- Byte Array
- List
- Dictionary/Map

On top of this, there also exist Graph data types (Node, Relationship, Path), temporal types and spatial types.

This unified type system is available throughout all the drivers, and all types are fully supported within Cypher (with the exception of byte arrays, which have no literal representation in the language). The list above should be the list from which you typically draw when building a Neo4j-backed application.

In terms of the other numeric types which have historically been available (e.g. 32-bit int, 16-bit int) these should be considered internal data types bound only to the low-level Java API. To that end, they are _only_ applicable for applications that use Neo4j in an embedded context. Bear in mind also that it is **strongly** recommended to use the same interface (e.g. embedded or Bolt) for both data input and output within the same application, to avoid data loss through type coercion.

Hopefully this clarifies that this is a deliberate design decision and not a bug, or a problem requiring a solution. Integer properties containing anything other than 64-bit signed integer values can _only_ have been created via a surface that will also allow extraction of those values. To reiterate: **do not mix surface APIs**.

Going forward, it is entirely possible that we may choose to expand on the type system available throughout the primary surface. This may be to introduce other numeric types. As and when that occurs, this will ripple throughout the Drivers and Cypher at the same time, since we aim to keep those type systems aligned.

One final suggestion: if you have a genuine need to store extended type information alongside your property values, there are techniques you can use to achieve this. One such approach is to store sibling properties that contain type information which your application can use for appropriate coercion.

**Comment by hvub** [comment]
No priority has been given to implement more integer types in Cypher in the last 5.5 years. Feel free to re-request the feature. Describing clear why exactly 16-bit or 32-bit integers are desired where 64-bit integer are insufficient, may increase the chance of these being considered.
